### DISCUSSION TITLE
Using the SD card and the TFT_eSPI driver, the system crashes or in other cases performs a hard reset on loading

### DISCUSSION THREAD

**Discussion by suntchi2018** [Bodmer/TFT_eSPI#2420] [General]
I apologize in advance, I'm a beginner with this type of product and I don't know exactly what data could be useful for you to understand this strange behavior.
If you need the whole program and a complete debug and the User_setup.h.

Everything works fine until I add the SD.h library, if I use SPIClass spi = SPIClass(VSPI); before setup().
The system does not finish compiling:
c:/users/admin/appdata/local/arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: libraries\TFT_eSPI\TFT_eSPI.cpp.o:C:\Users\admin\Documents\Arduino\libraries\TFT_eSPI\Processors/TFT_eSPI_ESP32.c:17: multiple definition of `spi'; sketch\Gestione_Stagionatore-ota.ino.cpp.o:C:\Users\admin\Documents\Arduino\Gestione_Stagionatore-ota/Gestione_Stagionatore-ota.ino:126: first defined here
collect2.exe: error: ld returned 1 exit status

If I don't put it, it compiles but ESP- WROOM 32 - 38 pin keeps resetting itself:
rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13836
load:0x40080400,len:3608
entry 0x400805f0
ets Jun  8 2016 00:22:57

The version of all libraries are updated to the latest version. I'm using Arduino Ide 1.8.19

These are the libraries in use

#include <Wire.h>
#include <FS.h>
#include <SD.h>
#include <SPI.h>
#include <TFT_eSPI.h>
#include <WiFi.h>
#include <WebServer.h>
//#include <ArduinoOTA.h>
#include "AiEsp32RotaryEncoder.h"
#include <BME280I2C.h>
#include <TimeLib.h>

I'm using the ST7789 external display similar to the one inserted on the TTGO D-display the dimensions are 1.14' RGB 135x240 - DRIVER IC:ST7789V 4WIRE SPI PORT
With the following ports GND-VCC-SCL-SDA-RES-DC-CS-BLK
I look forward to your response if possible.
![ESP32-WROOM-38-PIN](https://user-images.githubusercontent.com/40439798/219852616-4d993a67-39f7-47cb-941d-fd66559724d7.jpg)

**Comment by Bodmer** [comment]
The library creates an spi instance for the VSPI channel unless  #define USE_HSPI_PORT is in the setup file.

To create a different spi instance you need to use a different varaible name, for example:
```
SPIClass sdspi = SPIClass(VSPI);
    if(!SD.begin(22, sdspi)){
        Serial.println("Card Mount Failed");
        return;
    }
```

**Reply by proximacentaury** [reply]
That was my problem!
I have applied that solution with  different inctance. 
tft.getSPIinstance()
has changed to 
SPIClass sdspi = SPIClass(VSPI);  ...
(!SD.begin(5, sdspi))
Now the ESP32_SDcard_jpeg demo works with SUNTON CYD 2.4inch_ESP32-2432S024C version.

**Comment by suntchi2018** [comment]
Hi, thank you very much Bodmer for your help and your quick reply, but maybe I'm too incompetent, I tried to use pin 22 (GPIO8) for the SD card CS, but it still doesn't work, however the system does not crash, and this is a great improvement.
Where am I going wrong, if it's not too much trouble for your kindness.
This is the wiring diagram i used.
![Immagine](https://user-images.githubusercontent.com/40439798/219936328-4fdeccae-7d5e-4cda-befb-06d2582211ce.png)

**Reply by suntchi2018** [reply]
Ciao Bodmer,
nothing to do, I couldn't get the SD to work, it doesn't work I did a thousand tests but without results, he doesn't want to know. Leaving everything connected, rotary_encoder (which I need to manage the menu), the 6 relays, the SD and the display, if I use a test program for the SD it works, if I use a test program for rotary_encoder it works, if I try a program test to make the 6 relays work and the display always works without problems, if I use everything except the SD everything works, if I add the SD nothing works anymore.
I wanted to use the SD to avoid using the EEPROM, but I see that I'm not able to use it, I think it's just a dispute with the SD and honestly I'm not able to resolve.
It will mean that I will use the EEPROM, too bad!
Thanks anyway for your support.

**Comment by Bodmer** [comment]
See this example.
https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Generic/ESP32_SDcard_jpeg/ESP32_SDcard_jpeg.ino

Note that the chip selects are set high before the SD and TFT are initialised to avoid conflicts. Try that example.

**Comment by Bodmer** [comment]
I just realised you have the SD card on the same SPI pins, so creating a new SPI instance will cause problems.

The SPI instance must be the same as is used by TFT_eSPI, so this should work:
```
tft.init()
if(!SD.begin(22, tft.getSPIinstance())){
    Serial.println("Card Mount Failed");
    return;
}
```
It is also a good idea to set the chip selects high as the start of setup as shown in the ESP32_SDcard_jpeg example linked to above.

**Reply by suntchi2018** [reply]
Bodmer, thank you very kindly, but you are dealing with a naive, I can not solve the problem with this SD, I'm going crazy.
I tried using your solution but I can't get it to work.
I don't even want to waste your time you have already helped me so much and I thank you so much.
Unfortunately it's my fault, I pretend to do, but I don't have the experience I need in this environment.
I have to lower my expectations.
If you want, when I've finished it, I'll send you the project if you're interested in making salami and bresaola.
![TFT-SD-BME-encoder-relays](https://user-images.githubusercontent.com/40439798/220144029-80ab2f15-214f-4a72-9812-616da2bb5cd2.png)

**Comment by rradomir** [comment]
OH MY GOD, this solution solves my problem AT FIRST SHOT. Iam was struggling with this in few days. THANK YOU.
For the information: 
I have [Pico-Wave](https://www.waveshare.com/wiki/Pico-ResTouch-LCD-3.5) screen and RPi Pico. With this simple SD initialisation I get success status and simple example with new file "testing 1,2,3" WORKS.
I have ordered SDcard module in ten minutes before find this post.....
You definitely should include this info in main page :)
Thanks for GREAT library.
